tests/migrations/test_operations.py

```diff
@@ -179,6 +179,835 @@ def test_create_model_with_duplicate_manager_name(self):
+                )
+                fk_fieldname = f"pony_{on_delete_type}"
+                fk_parent_model = f"Pony_{on_delete_type}"
+                fk_child_model = f"Rider_{on_delete_type}"
+
+                Rider = project_state.apps.get_model(app_label, fk_child_model)
+                self.assertEqual(
+                    Rider._meta.get_field(fk_fieldname).remote_field.on_delete,
+                    non_db_cascade_options[on_delete_type][0],
+                )
+                # Test migrations for db to non db changes
+                # Test with db set null
+                self.change_to_db_set_null_test(
+                    app_label,
+                    fk_parent_model,
+                    fk_child_model,
+                    fk_fieldname,
+                    project_state,
+                )
+                self.database_backward_assertion(
+                    app_label,
+                    project_state,
+                    on_delete_type,
+                    fk_parent_model,
+                    fk_child_model,
+                    fk_fieldname,
+                )
+                # Test with db restrict
+                self.change_to_db_restrict_test(
+                    app_label,
+                    fk_parent_model,
+                    fk_child_model,
+                    fk_fieldname,
+                    project_state,
+                )
+                self.database_backward_assertion(
+                    app_label,
+                    project_state,
+                    on_delete_type,
+                    fk_parent_model,
+                    fk_child_model,
+                    fk_fieldname,
+                )
+                # Test with db cascade
+                self.change_to_db_cascade_test(
+                    app_label,
+                    fk_parent_model,
+                    fk_child_model,
+                    fk_fieldname,
+                    project_state,
+                )
+                self.database_backward_assertion(
+                    app_label,
+                    project_state,
+                    on_delete_type,
+                    fk_parent_model,
+                    fk_child_model,
+                    fk_fieldname,
+                )
+                # Test with db_set default
+                if connection.features.has_on_delete_db_default:
+                    self.change_to_db_set_default_test(
+                        app_label,
+                        fk_parent_model,
+                        fk_child_model,
+                        fk_fieldname,
+                        project_state,
+                    )
+                    self.database_backward_assertion(
+                        app_label,
+                        project_state,
+                        on_delete_type,
+                        fk_parent_model,
+                        fk_child_model,
+                        fk_fieldname,
+                    )
```

Comment:
The number of nested hooks makes these tests/assertions extremely difficult to understand (at least for me).

Reply:
This seems to be a persistent issue, will go forward with the suggestion made by Lily [here](https://github.com/django/django/pull/16851#issuecomment-1755491297). Let me know if you approve of creating a new file for this, as there is going to be a lot of tests. Will start after few days.
>If you can't get that working, I think giving up on subTest might be reasonable - maybe create a new testcase or test file for these tests and just have all of them as distinct test methods? I know there's going to be loads of them, but maybe that's easier to write...